File: api/render.js
import fs from "fs/promises";
import path from "path";
import { fileURLToPath } from "url";

const filename = fileURLToPath(import.meta.url);
const dirname = path.dirname(filename);

export default {
  async fetch(request) {
    try {
      if (request.method != "GET") {
        return new Response(null, { status: 405 });
      }

      const url = new URL(request.url, `http://${request.headers.host}`);
      const url_arr = url.pathname.split("/").slice(1);

      let page = url_arr[0];

      if (page == "" || page == "index") {
        page = "index";
      }

      const layoutPath = path.join(dirname, "..", "templates", "layout.html");
      const pagePath = path.join(dirname, "..", "pages", `${page}.html`);

      const [layout, pageContent] = await Promise.all([
        fs.readFile(layoutPath, "utf-8"),
        fs.readFile(pagePath, "utf-8"),
      ]);

      const html = layout
        .replace("{content}", pageContent)
        .replace("{pathname}", url.pathname)
        .replace("{bytes}", pageContent.length);

      return new Response(html, {
        status: 200,
        headers: {
          "content-type": "text/html; charset=utf-8",
        },
      });
    } catch (err) {
      return new Response(err.message, { status: 500 });
    }
  },
};

File: api/render.rs
use vercel_runtime::{Error, Request, Response, ResponseBody, run, service_fn};

use website::app;

#[tokio::main]
async fn main() -> Result<(), Error> {
    let service = service_fn(handler);
    run(service).await
}

async fn handler(req: Request) -> Result<Response<ResponseBody>, Error> {
    if req.method() != "GET" {
        return Ok(Response::builder()
            .status(405)
            .body(ResponseBody::from(()))?);
    }

    let url_pathname = req.uri().path();
    let url_arr: Vec<&str> = url_pathname.split("/").skip(1).collect();
    let page_name = if !url_arr.is_empty() && !url_arr[0].is_empty() {
        url_arr[0]
    } else {
        "index"
    };

    let page = match page_name {
        "index" => app::pages::index::page(),
        "social" => app::pages::social::page(),
        "dev" => app::pages::dev::page(),
        _ => app::pages::not_found::page(),
    }
    .into_string();

    let layout = app::layout::page(&page, url_pathname);
    let html = layout.into_string();

    Ok(Response::builder()
        .status(200)
        .header("Content-Type", "text/html")
        .body(ResponseBody::from(html))?)
}
